replication-api-impl/src/main/java/com/connexta/replication/api/impl/worker/Worker.java

```diff
@@ -0,0 +1,259 @@
+        doCreate(task, metadata, destination, local);
+        break;
+      case REPLICATE_TO:
+        doCreate(task, metadata, local, destination);
+        break;
+      case DELETE:
+        // always delete on non-local site
+        doDelete(task, metadata);
+        break;
+      default:
+        // should never hit this since deserialization will return UNKNOWN
+        task.fail(ErrorCode.UNSUPPORTED_ERROR,
+            String.format("Operation task %s is not supported by this worker", task));
+    }
+  }
+
+  private void doCreate(final Task task, final Metadata metadata, final NodeAdapter source,
+      final NodeAdapter destination)
+      throws InterruptedException {
+    final String metadataId = task.getId();
+    final String operation = task.getOperation().name();
+    if (source.exists(metadata)) {
+      CreateRequest request = new CreateRequestImpl(List.of(metadata));
+      boolean success = destination.createRequest(request);
+      if (!success) {
+        task.fail(ErrorCode.OPERATION_FAILURE,
+            String.format("Failed to %s metadata %s from %s to %s",
+                operation,
+                metadataId,
+                source.getSystemName(),
+                destination.getSystemName()));
+      }
+    } else {
+      LOGGER.info("Received task to create metadata on site {}, but it was deleted on {}",
+          destination.getSystemName(), destination.getSystemName());
+      task.complete();
+    }
+
+    ResourceInfo resourceInfo = task.getResource().orElse(null);
+    if (resourceInfo != null) {
+      URI uri = resourceInfo.getResourceUri().orElse(null);
+      if (uri != null) {
+        ResourceRequest request = new ResourceRequestImpl(metadata);
+        try {
+          ResourceResponse response = source.readResource(request);
+          Resource resource = response.getResource();
+          boolean success = destination
+              .createResource(new CreateStorageRequestImpl(List.of(resource)));
+          if (!success) {
+            task.fail(ErrorCode.OPERATION_FAILURE,
+                String.format("Failed to %s resource %s from %s to %s",
+                    operation,
+                    uri,
+                    source.getSystemName(),
+                    destination.getSystemName()));
+          }
+        } catch (AdapterException e) {
+          LOGGER.debug("Failed to retrieve resource", e);
+          task.fail(ErrorCode.OPERATION_FAILURE,
+              String.format("Failed to read resource from %s", destination.getSystemName()));
+        }
+      } else {
+        LOGGER.warn("No resource URI specified for task {}", task);
+        return;
+      }
+    }
+    task.complete();
+  }
+
+  private void doDelete(Task task, Metadata metadata) throws InterruptedException {
+    final String metadataId = task.getId();
+    final String operation = task.getOperation().name();
+    boolean success = destination.deleteRequest(new DeleteRequestImpl(List.of(metadata)));
```

Comment:
❗️ Need to handle adapter exception.
❓ What happens if it doesn't exist on the site? Can we detect that type of failure from the adapter and throw a different error so that here we do not have to fail since it is a no-op and therefore the task should be completed.

Reply:
Caught adapter exception and for now did an exists check and left a todo to remove once adapters support better error reporting